### ISSUE TITLE
[apex] ExcessiveClassLength multiple warning on the same class

### ISSUE THREAD

**Issue by stefanoportelli** [pmd/pmd#3142]
**Affects PMD Version**: 6.31

**Rule:** [ExcessiveClassLength](https://pmd.github.io/latest/pmd_rules_apex_design.html#excessiveclasslength)

**Description:**
With an apex class that contains more than 9300 lines and more than 700 public class definition PMD fires ono warning for each public class definition instead of just one warning.

The class is something like that:

```java
public class Foo {
    public class bar1() {
        // about 50 lines of code
    }
    public class bar2() {
        // about 50 lines of code
    }
    public class bar3() {
        // about 50 lines of code
    }
.....
.....
    public class barN() {
        // about 50 lines of code
    }
}
```

**Running PMD through:** *CLI*

**Comment by rsoesemann** [comment]
9300 lines and more than 700 public class definition PMD ???? Thank god that PMD doesn't explode ! Just kidding.

**Comment by adangel** [comment]
I can reproduce it with the following sample and setting threshold to 10 lines:

```java
public class FooSmaller {

    public class bar1 {
        public void m() {
            System.out.println('foo');
            System.out.println('foo');
        }
    }

    public class bar2 {
        public void m() {
            System.out.println('foo');
            System.out.println('foo');
        }
    }
}
```

Here we expect one violation for the outer class, but the inner classes are both < 10 lines, so shouldn't be reported:

```
src/FooSmaller.cls:1:	ExcessiveClassLength:	Avoid really long classes.
src/FooSmaller.cls:3:	ExcessiveClassLength:	Avoid really long classes.
```

It seems, we don't have the locations correct for the inner classes. In the designer, I see, that class "bar1" spans from line 3 to 16, "bar2" from line 10 to 16 and "FooSmaller" from 1 to 16.

Note: Your original example doesn't compile is not parsable and not valid apex code.

**Comment by stefanoportelli** [comment]
Hi @adangel, just to have a better overview of the issue and a parsable and valid apex code :) check the attache file.

Using PMD, it reports 87 E**xcessiveClassLength** warnings on this class.
The first on line 3: public class MetadataService {
Then line 4:     public class listMetadataResponse_element {
Line 10:     public class WorkflowRule extends Metadata {
...
is that an expected behavior? 

Thanks in advance,
Stefano
[sampleclass.txt](https://github.com/pmd/pmd/files/6089302/sampleclass.txt)

**Comment by adangel** [comment]
Hi @stefanoportelli ,
I assume, you use the standard configuration of the rule, which is: report if more than 1000 lines.
That means, the expected warnings are:

1. for the outer class "public class MetadataService {" - line 3

and that's all. There are some bigger inner classes (like CustomField), but they are not bigger than 1000 lines (e.g. this one "only" 100 lines).

I can confirm, that with the bug fix (#3156) only this one violation is reported.

> is that an expected behavior?

In short: no.

I've seen, that this class is generated ("Generated by wsdl2apex"). If it is not subject to manual modification, you might also exclude this file from analysis. There are different opinions, whether generated sources should be checked or not....
